File: 03day/part1.hs
import Data.Char (isAsciiLower, ord)
import Data.Set (fromList, intersection, toList, union)
import System.IO (IOMode (ReadMode), hClose, hGetContents, openFile)

score :: Char -> Int
score c
  | isAsciiLower c = (ord c - ord 'a') + 1
  | otherwise = (ord c - ord 'A') + 27

solve :: String -> Int
solve xs = sum $ map score $ toList $ a `intersection` b
  where
    half = length xs `div` 2
    [a, b] = map fromList $ (\(a, b) -> [a, b]) $ splitAt half xs

solution :: String -> Int
solution = sum . map solve . lines

main :: IO ()
main = do
  handle <- openFile "input.txt" ReadMode
  content <- hGetContents handle
  print $ solution content
  hClose handle

File: 03day/part1.py
def solve(line):
    L = len(line)
    first, second = set(line[:L // 2]), set(line[L // 2:])
    similar = first & second
    tot = 0
    for item in similar:
        if ord('a') <= ord(item) <= ord('z'):
            tot += (ord(item) - ord('a')) + 1
        else:
            tot += (ord(item) - ord('A')) + 27
    return tot


if __name__ == "__main__":
    lines = []
    with open("input.txt") as f:
        for line in f:
            lines.append(line.strip())
    print(sum(map(solve, lines)))
